src/envoy/http/backend_routing/filter.cc

```diff
@@ -52,50 +53,74 @@ FilterHeadersStatus Filter::decodeHeaders(
-              operation, newPath);
-  } else {  // APPEND_PATH_TO_ADDRESS
-    newPath = absl::StrCat(rule->path_prefix(), original_path);
-    config_->stats().append_path_to_address_request_.inc();
-    ENVOY_LOG(
-        debug,
-        "append path to address backend routing for operation {}, new path: {}",
-        operation, newPath);
+
+  absl::string_view original_path = headers.Path()->value().getStringView();
+  std::string new_path;
+
+  switch (rule->path_translation()) {
+    case BackendRoutingRule::CONSTANT_ADDRESS:
+      new_path = translateConstPath(rule, original_path);
+      config_->stats().constant_address_request_.inc();
+      ENVOY_LOG(debug,
+                "constant address backend routing for operation {}"
+                ", original path: {}, new path: {}",
+                operation, original_path, new_path);
+      break;
+
+    case BackendRoutingRule::APPEND_PATH_TO_ADDRESS:
+      new_path = translateAppendPath(rule, original_path);
+      config_->stats().append_path_to_address_request_.inc();
+      ENVOY_LOG(debug,
+                "append path to address backend routing for operation {}"
+                ", original path: {}, new path: {}",
+                operation, original_path, new_path);
+      break;
+
+    default:
+      NOT_REACHED_GCOVR_EXCL_LINE;
   }
-  const auto& pathField = Envoy::Http::Headers::get().Path;
-  headers.remove(pathField);
-  headers.addCopy(pathField, newPath);
 
+  headers.setPath(new_path);
   return FilterHeadersStatus::Continue;
 }
 
+std::string Filter::translateConstPath(const BackendRoutingRule* rule,
+                                       absl::string_view original_path) {
+  const auto& filter_state = *decoder_callbacks_->streamInfo().filterState();
+  absl::string_view extracted_query_params =
+      utils::getStringFilterState(filter_state, utils::kQueryParams);
+
+  const auto original_path_str = std::string(original_path);
+  std::size_t originalQueryParamPos = original_path_str.find('?');
+  if (originalQueryParamPos == std::string::npos) {
+    // No query param in original request.
+    std::string new_path = rule->path_prefix();
+    if (!extracted_query_params.empty()) {
+      // Add extracted variable bindings.
+      absl::StrAppend(&new_path, "?", extracted_query_params);
+    }
+    return new_path;
+  }
+
+  // Has query parameters in original request.
+  const std::string& originalQueryParam =
+      original_path_str.substr(originalQueryParamPos);
+  std::string new_path = absl::StrCat(rule->path_prefix(), originalQueryParam);
+  if (!extracted_query_params.empty()) {
+    // Append extracted variable bindings.
+    absl::StrAppend(&new_path, "&", extracted_query_params);
+  }
+  return new_path;
+}
+
+std::string Filter::translateAppendPath(const BackendRoutingRule* rule,
+                                        absl::string_view original_path) {
+  // Just append the original request path to the configured path prefix.
```

Comment:
Put the comments in front of function name?

Reply:
Done